go/database/mpt/forest_test.go

```diff
@@ -2155,6 +2155,147 @@ func TestForest_VisitPathToStorage(t *testing.T) {
+					directory := t.TempDir()
+					var err error
+					forest, err := variant.factory(directory, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					root := NewNodeReference(EmptyId())
+
+					addresses := getTestAddresses(N)
+					for i, address := range addresses {
+						root, err = forest.SetAccountInfo(&root, address, AccountInfo{Balance: common.Balance{0x1}})
+						if err != nil {
+							t.Fatalf("cannot update account: %v", err)
+						}
+						// fill storage of half of the accounts
+						if i < 50 {
+							continue
+						}
+						root, err = forest.SetValue(&root, address, common.Key{byte(i)}, common.Value{byte(i)})
+						if err != nil {
+							t.Fatalf("cannot update storage: %v", err)
+						}
+					}
+
+					for i := 0; i < N; i++ {
+						// First half of the accounts has empty storage
+						if i < 50 {
+							isEmpty, err := forest.HasEmptyStorage(&root, addresses[i])
+							if err != nil {
+								t.Fatalf("failed to ask whether account has empty storage: %v", err)
+							}
+							if !isEmpty {
+								t.Error("freshly created account has empty storage, but forest returned true")
+							}
+							continue
+						}
+
+						// Second half of the accounts has non-empty storage
+						if i >= 50 {
+							isEmpty, err := forest.HasEmptyStorage(&root, addresses[i])
+							if err != nil {
+								t.Fatalf("failed to ask whether account has empty storage: %v", err)
+							}
+							if isEmpty {
+								t.Error("storage is not empty but forest returned true")
+							}
+							continue
+						}
+
+						// Clear some storage
+						if i >= 60 && i < 70 {
+							// Clear storage of an account with not empty storage
+							root, err = forest.ClearStorage(&root, addresses[i])
+							if err != nil {
+								t.Fatalf("cannot clear storage: %v", err)
+							}
+
+							// Account with cleared storage
+							isEmpty, err := forest.HasEmptyStorage(&root, addresses[i])
+							if err != nil {
+								t.Fatalf("failed to ask whether account has empty storage: %v", err)
+							}
+							if !isEmpty {
+								t.Error("storage was cleared but forest returned false")
+							}
+							continue
+						}
+
+						// Delete some accounts
+						if i >= 70 && i < 80 {
+							root, err = forest.SetAccountInfo(&root, addresses[i], AccountInfo{})
+							if err != nil {
+								t.Fatalf("failed to set account info: %v", err)
+							}
+							isEmpty, err := forest.HasEmptyStorage(&root, addresses[i])
+							if err != nil {
+								t.Fatalf("failed to ask whether account has empty storage: %v", err)
+							}
+							if !isEmpty {
+								t.Error("deleted account has empty storage forest returned false")
+							}
+							continue
+						}
+
+						// Set empty storage for the rest
+						root, err = forest.SetValue(&root, addresses[i], common.Key{byte(i)}, common.Value{})
```

Comment:
I do not understand why all addresses that did not pass previous conditions are deleted here anyway. If the all possible options were already checked? What is being checked here?

Reply:
First condition checks that the storage has actually been created.
Second condition calls `ClearStorage` storage for addreses between index 60 and 70 and checks that `HasEmptyStorage` returns true
Third condition deletes the account etween index 70 and 80 and checks that `HasEmptyStorage` returns true.
The last code snipet takes the rest of the indexes hence <=80  and calls `SetValue` with an empty value to remove the storage and checks that `HasEmptyStorage` returns true.